README.md
# PostCSS Maze

A fully flexible mobile first grid to suit any design pattern.

[PostCSS]: https://github.com/postcss/postcss
[ci-img]:  https://travis-ci.org/cathydutton/postcss-maze.svg
[ci]:      https://travis-ci.org/cathydutton/postcss-maze


Maze is a flexible and semantic mobile first responsive grid built with PostCSS. Maze is fully customisable and removes the reliance on .last-column classes by instead applying the margin to both sides of each element and adjusting the wrapper width accordingly.


## Installation

```js
npm install postcss-maze
```


## Creating a grid

Firstly the wrapper is given a property of grid-container with a value of true.

```css
.wrapper {
  grid-container: true;
}
```

This calculates the wrapper width based on the values assigned to the largest and smallest media queries, and adds a clearfix.

The max width is calculated accounting for the element margins...

```css
.wrapper {
  max-width: 1030.2px;
  min-width: 260px;
  margin: 0 auto;
  box-sizing: border-box;
}

.wrapper::after {
  content: "" ;
  display: table;
  clear: both;
}
```
Each column is then created as a ratio at a set media query...

```css
.four-elements {
  col-span: mobile(1,1);
  col-span: phablet(1,2);
  col-span: tablet(1,4);
}
```

```css
.four-elements {
  float: left;
  box-sizing: border-box;
  margin-right: 0.5%;
  margin-left: 0.5%;
  width: 99%;
}

@media only screen and (min-width:30em) {

  .four-elements {
    width: 49%;
  }

}

@media only screen and (min-width:48em) {

  .four-elements {
    width: 24%;
  }

}
```

The above element will display one per row at mobile, two per row at landscape and 4 per row at tablet. The Mobile declaration is the default value and therefore is not rendered inside a media query.

### Examples

```css
.twelve-elements {
  col-span: mobile(1,1);
  col-span: phablet(1,2);
  col-span: tablet(1,4);
  col-span: desktop(1,6);
  col-span: wide(1,12);
}
```

```css
.six-elements {
  col-span: mobile(1,2);
  col-span: tablet(1,3);
  col-span: desktop(1,6);
}
```

### Settings

By default Maze works with 5 media queries and has a margin of 1%. These settings can be overridden with custom config.

All media query dimensions should be written in EM's, and margins as a percentage...

```js
var processors = [
postcssmaze({
media: {
 mobile:    20,
 phablet:   30,
 tablet:    48,
 desktop:   63.750,
 wide:      80
},
settings: {
 margin: 10
}
}),
```

An example using Gulp as a task runner...

```js
// CSS TASK
gulp.task('css', function () {
  var concat = require('gulp-concat'),
    postcss = require('gulp-postcss'),
    mqpacker = require('css-mqpacker'),
    postcssmaze = require('postcss-maze'),
    autoprefixer = require('autoprefixer');

  var processors = [
    postcssmaze({
      media: {
        mobile:    20,
        phablet:   30,
        tablet:    48,
        desktop:   63.750,
        wide:      80
      },
      settings: {
        margin: 10
      }
    }),
    mqpacker,
      autoprefixer({
        browsers: ['> 2%', 'IE >= 9', 'iOS >= 7'],
        cascade:  false,
        map:      true,
        remove:   true
    })
  ];

  return gulp.src('src/input.css')
    .pipe(postcss(processors))
    .pipe(concat('dist/output.css'))
    .pipe(gulp.dest('.'));
});


```

### Support
If you have any questions get in touch:

-   [info@cathydutton.co.uk](mailto:info@cathydutton.co.uk)
-   [@cathy_dutton](http://twitter.com/cathy_dutton)


CHANGELOG.md
# 1.0.0 - 2016-02-22
* Initial release


test/test.js
var postcss = require('postcss'),
    grid = require('../'),
    assert = require('assert');

var settingsConfig = {
    margin: 1
};

var test = function (opts, input, output) {
    assert.equal(postcss(grid(opts)).process(input).css, output);
};

test(settingsConfig, '.element{col-span: mobile(1,1);}', '.element{float: left;box-sizing: border-box;margin-right: 0.5%;margin-left: 0.5%;width: 99%;transition: all 0.5s ease;}');
test(settingsConfig, '.element{col-span: mobile(1,2);}', '.element{float: left;box-sizing: border-box;margin-right: 0.5%;margin-left: 0.5%;width: 49%;transition: all 0.5s ease;}');
test(settingsConfig, '.element{col-span: mobile(1,3);}', '.element{float: left;box-sizing: border-box;margin-right: 0.5%;margin-left: 0.5%;width: 32.333333333333336%;transition: all 0.5s ease;}');


gulpfile.js
var gulp = require('gulp');

var files = ['index.js', 'test/*.js', 'gulpfile.js'];

gulp.task('lint', function () {
    var eslint = require('gulp-eslint');
    return gulp.src(files)
        .pipe(eslint())
        .pipe(eslint.format())
        .pipe(eslint.failAfterError());
});

// CSS TASK
gulp.task('css', function () {
    var concat = require('gulp-concat'),
        postcss = require('gulp-postcss'),
        mqpacker = require('css-mqpacker'),
        postcssmaze = require('./index.js');

    var processors = [
        postcssmaze({
            media: {
              // mobile:    20,
              // phablet:   30,
              // tablet:    48,
              // desktop:   63.750,
              // wide:      80
            },
            settings: {
                // margin: 10
            }
        }),
        mqpacker
    ];

    return gulp.src([
        'demo/layout.css',
        'demo/theme.css'
    ])
        .pipe(postcss(processors))
        .pipe(concat('demo/output.css'))
        .pipe(gulp.dest('.'));
});

gulp.task('test', function () {
    var mocha = require('gulp-mocha');
    return gulp.src('test/*.js', {
        read: false
    })
      .pipe(mocha());
});

gulp.task('default', ['lint', 'test']);

gulp.task('watch', function () {
    gulp.watch(files, ['css', 'lint']);
});

gulp.task('deploy', function () {
    var ghPages = require('gulp-gh-pages');
    return gulp.src('./demo/**/*')
    .pipe(ghPages());
});


index.js
// max width - exact value -- rounding
// Write test - lint test.JS

var postcss = require('postcss'),
    _ = require('underscore');

// Default Media Config
var mediaConfig = {
    mobile:  20,
    phablet: 30,
    tablet:  48,
    desktop: 63.750,
    wide:    80
};

// Default Settings Config
var settingsConfig = {
    margin: 1
};

module.exports = postcss.plugin('postcss-maze', function (options) {

    // Extend the default Configs with any custom values.
    options = options || {};
    mediaConfig = _.extend(mediaConfig, options.media);
    settingsConfig = _.extend(settingsConfig, options.settings);

    // Width Function
    var columnWidth = function (coloumnSpan, totalColoumns) {
        var unitWidth = totalColoumns / coloumnSpan;
        var width = (100 - unitWidth * settingsConfig.margin) / unitWidth;
        return width + '%';
    };

    // Set widths - Add to the max width to compensate for margin
    var maxWidth = mediaConfig.wide + mediaConfig.wide / 100 *
      settingsConfig.margin;
    var minWidth = mediaConfig.mobile; /* Set min width with scroll bar */


    return function (css) {
        css.walkRules(function (rule) {
            rule.walkDecls(function (decl) {

                var value = decl.value,
                    selectorName = decl.parent.selector,
                    prop = decl.prop;

                // Set up grid container
                if (prop === 'grid-container') {
                    var gridContainerCheck = /^\s*true\s*$/;

                    if (!value.match(gridContainerCheck)) {
                        throw decl.error('Syntax Error - grid-container:true', {
                            plugin: 'postcss-maze'
                        });
                    } else {
                        decl.parent.append({
                            prop:  'max-width',
                            value: maxWidth + 'em'
                        });
                        decl.parent.append({
                            prop:  'min-width',
                            value: 'calc(' + minWidth + 'em - 20px)'
                        });
                        decl.parent.append({
                            prop:  'margin',
                            value: '0 auto'
                        });
                        decl.parent.append({
                            prop:  'box-sizing',
                            value: 'border-box'
                        });
                        css.insertBefore(
                          decl.parent,
                            selectorName + '::after {' +
                            'content: "" ;' +
                            ' display: table;' +
                            'clear: both }'
                          );
                    }
                    decl.remove();
                }

                // Set up grid elements
                if (prop === 'col-span') {
                    var colCheck = /^([a-z]+)(\([0-9]{1,2})(,[0-9]{1,2}\))$/;

                    if (!value.match(colCheck)) {
                        throw decl.error('Syntax Error - media(value,value)', {
                            plugin: 'postcss-maze'
                        });
                    } else {

                        // Variables
                        var split = value.split('('); // Split Media from values
                        // Media (mobile, desktop etc)
                        var chosenMedia = split[0];
                        var values = split[1]; // Grid values
                        // Remove trailing ) and split the two values (2,6)
                        var splitValues = values.slice(0, -1).split(',');
                        var span = splitValues[0]; // Col Span
                        var coloumns = splitValues[1]; // Total coloumns
                        var mediaValue = mediaConfig[chosenMedia];
                        var marginValue = settingsConfig.margin;

                        if (chosenMedia === 'mobile') {
                            decl.parent.append({
                                prop:  'float',
                                value: 'left'
                            });
                            decl.parent.append({
                                prop:  'box-sizing',
                                value: 'border-box'
                            });
                            decl.parent.append({
                                prop:  'margin-right',
                                value: marginValue / 2 + '%'
                            });
                            decl.parent.append({
                                prop:  'margin-left',
                                value: marginValue / 2 + '%'
                            });
                            decl.parent.append({
                                prop:  'width',
                                value: columnWidth(span, coloumns)
                            });
                            decl.parent.append({
                                prop:  'transition',
                                value: 'all 0.5s ease'
                            });
                        } else {
                            // add media query
                            css.insertBefore(decl.parent,
                              '@media only screen and (min-width:' +
                              mediaValue +
                              'em) { ' +
                              selectorName +
                              '{width:' +
                              columnWidth(span, coloumns) +
                              '} }');
                        }

                    }

                    decl.remove();
                }

            });

        });

    };

});
